### PR TITLE
feat: Payroll based on attendance or leave

### PR DESCRIPTION
**Pull request by Anurag810**
We can set the payroll based on **attendance/Leave** from **HR settings**. As shown below.

HR SETTINGS

Salary slip will consider:
 
- **Leave without pay** as non-working days, if payroll based on is set to **Leave**.
-  **Absent days** as non-working days, if payroll based on is set to Attendance and it will consider attendance record with status **Absent** and **Half day**. The **half-day** is considered **half-day absent** during the creation of a salary slip.

Please suggest any enhancement required for this feature.
And can we provide this feature for multiple companies?

### CODE REVIEW

**Comment by nabinhait** [comment]
@Anurag810 please resolve conflict

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 1]
```diff
@@ -193,21 +197,37 @@ def get_leave_details(self, joining_date=None, relieving_date=None, lwp=None, fo
...
 
 		holidays = self.get_holidays_for_employee(self.start_date, self.end_date)
-		actual_lwp = self.calculate_lwp(holidays, working_days)
+		non_working_days = 0
+		if payroll_based_on == 'Leave':
+			actual_lwp = self.calculate_lwp(holidays, working_days)
+		elif payroll_based_on == 'Attendance':
+			actual_absent_days = self.calculate_absents()
+		else:
+			frappe.throw(_("Please set <b>Payroll based on</b> in <b>HR settings</b>"))
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: Don't use HTML tags inside message, it creates problem in translation

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 2]
```diff
@@ -272,6 +292,25 @@ def calculate_lwp(self, holidays, working_days):
 				lwp = cint(leave[0][1]) and (lwp + 0.5) or (lwp + 1)
 		return lwp
 
+	def calculate_absents(self):
+		absent_days = 0
+
+		attendances = frappe.db.sql('''SELECT status, count(*) as non_working_days
+			FROM `tabAttendance`
+			WHERE
+				status NOT IN ("Present","On leave")
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py:308]: "On leave" with leave_type as LWP should also be treated as absent.

File: erpnext/hr/doctype/salary_slip/salary_slip.py [hunk 3]
```diff
@@ -272,6 +292,25 @@ def calculate_lwp(self, holidays, working_days):
...
+				AND employee = "{0}"
```
Comment (nabinhait) [inline erpnext/hr/doctype/salary_slip/salary_slip.py]: This will give error if employee value has '"', instead write like `AND employee = %s`

**Review by nabinhait — CHANGES_REQUESTED** [review]
- Add docs
- Add tests
- Write a patch to set `payroll_based_on` as 'Leave' in existing account.

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 1 above)
Comment (Anurag810) [reply erpnext/hr/doctype/salary_slip/salary_slip.py]: Done

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 2 above)
Comment (Anurag810) [reply erpnext/hr/doctype/salary_slip/salary_slip.py:308]: Done

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 3 above)
Comment (Anurag810) [reply erpnext/hr/doctype/salary_slip/salary_slip.py]: Done

**Comment by Olawale1** [comment]
Hi @Anurag810 

Thanks a lot for taking this up. I think it's important to add a field in HR Settings where the user can specify the fraction to be used for Half Day attendance. This is simply because many companies use the Half Day attendance to mark lateness to work and may not necessarily want that to deduct a half day's wages. The value could however be set as 0.5 by default

Cheers

File: erpnext/hr/doctype/salary_slip/salary_slip.py (hunk 2 above)
Comment (nabinhait) [reply erpnext/hr/doctype/salary_slip/salary_slip.py:308]: "On leave" with leave_type as LWP?

**Review by nabinhait — CHANGES_REQUESTED** [review]

**Comment by nabinhait** [comment]
> Hi @Anurag810
> 
> Thanks a lot for taking this up. I think it's important to add a field in HR Settings where the user can specify the fraction to be used for Half Day attendance. This is simply because many companies use the Half Day attendance to mark lateness to work and may not necessarily want that to deduct a half day's wages. The value could however be set as 0.5 by default
> 
> Cheers

To mark late attendance, you should use a custom status or a custom field altogether. If you mix it up, then it can cause error in many places.

**Comment by Olawale1** [comment]
> > Hi @Anurag810
> > Thanks a lot for taking this up. I think it's important to add a field in HR Settings where the user can specify the fraction to be used for Half Day attendance. This is simply because many companies use the Half Day attendance to mark lateness to work and may not necessarily want that to deduct a half day's wages. The value could however be set as 0.5 by default
> > Cheers
> 
> To mark late attendance, you should use a custom status or a custom field altogether. If you mix it up, then it can cause error in many places.

Hi @nabinhait 

I don't see how this will cause any errors anywhere else. What I am suggesting is for this feature only and I think it's quite critical too. Maybe I should try to explain a little better:

With this feature in it's current state, let's assume there are 20 Working Days and the employee's salary is $2000. If he/she is marked as Present for 18 Days and Half Day for 2 Days, the system will calculate salary as ((18+0.5+0.5)/20) x 2000 = $1900

The issue however is that Half Day in Attendance doesn't always mean that the Employee worked for exactly half of the working hours. It is mostly generically used to show that the Employee did not work full hours (lateness) on that day and hence, some organizations do not penalize the Employee for exactly half of the day's wages

Using my suggestion above, If Payroll based on 'Attendance' is selected in HR Settings, then an additional field appears to confirm fraction (or percentage) of Daily Wages to be deducted for Half Day attendance (0.5 or 50% by default). If the user adjusts this value to let's say 25%, the salary of the Employee in the earlier example is then calculated as ((18+0.75+0.75)/20) x 2000 = $1950

It's just a simple tweak in the calculation (factoring in the percentage specified for Half Day penalty), that's all! Using custom fields on the other hand would be much more complex and will not work with the feature unless the core code is hacked!

Kindly consider this

Thanks

**Comment by Anurag810** [comment]
@nabinhait please merge

**Comment by michellealva** [comment]
@Anurag810 what do we do if we want payroll to be based on both attendance and leave?

**Comment by Anurag810** [comment]
> @Anurag810 what do we do if we want payroll to be based on both attendance and leave?
@michellealva 
There is an option to set attendance/leave in HR settings.

**Comment by Anurag810** [comment]
Added tests

**Comment by fnrfarid** [comment]
Hi,

I would like to suggest to develop an option/automation that 5 [x] days late should be considered as an absent or leave without pay. Also, there should be some configurable slots like, for 5 days, 1 LWP, for 6-9 days, 2 days LWP, for 10-14 days, 3 days LWP etc. 

Late attendances should result as absent or leave without pay as many companies follow this policy.

**Comment by fnrfarid** [comment]
> 
> 
> @Anurag810 what do we do if we want payroll to be based on both attendance and leave?

@Anurag810 this is actually a valid requirement. I would like to process the payroll based on attendance, as well as based on leave without pay too. I guess, if we select "based on attendance" then, other leave without pay could be managed manually! then again, it would be a hassle.

**Comment by Olawale1** [comment]
> I would like to suggest to develop an option/automation that 5 [x] days late should be considered as an absent or leave without pay. Also, there should be some configurable slots like, for 5 days, 1 LWP, for 6-9 days, 2 days LWP, for 10-14 days, 3 days LWP etc.
> 
This just buttresses my point about not making the Half Day attendance a rigid 50% Daily Wage deduction. It needs to be a variable set by the user

Kind regards,

**Comment by Muad96** [comment]
I'd suggest to make **Absent Days** as Salary Component [Deduction] than can be configured from HR Settings. This is to avoid confusion when auditing/trace back the **LWP** days when **LWP** calculate from both **Leave Application** & **Attendance** Doctype.

**Comment by Anurag810** [comment]
@nabinhait please review and merge

**Comment by fnrfarid** [comment]
@anurag810, could you kindly share the latest updates on this?

**Comment by nabinhait** [comment]
Create another refactored PR, closing this.
https://github.com/frappe/erpnext/pull/21258

**Pull request closed without merging**